spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRowLineagePropagation.java

```diff
@@ -0,0 +1,489 @@
+
+    List<Record> expectedRecords = Lists.newArrayList();
+    expectedRecords.addAll(
+        generatedRecords.stream()
+            .filter(record -> record.get(0, Integer.class) != 1)
+            .collect(Collectors.toList()));
+    Record updatedRecord = createRecord(schema, 5678, 1L, null);
+    expectedRecords.add(updatedRecord);
+    table.refresh();
+
+    assertRecordsAreEqual(sortedRecordsByRowId(expectedRecords), recordsSortedByRowId(table));
+  }
+
+  protected Record createRecord(Schema schema, int id, Long rowId, Long lastUpdatedSequenceNumber) {
+    Record record = GenericRecord.create(schema);
+    record.set(0, id);
+    record.set(1, rowId);
+    record.set(2, lastUpdatedSequenceNumber);
+    return record;
+  }
+
+  protected List<Record> recordsSortedByRowId(Table table) throws IOException {
+    Schema schema = TypeUtil.join(table.schema(), rowLineageSchema);
+    List<Record> records = Lists.newArrayList();
+    table.refresh();
+    Snapshot snapshot = table.currentSnapshot();
+    if (branch != null) {
+      snapshot = table.snapshot(branch);
+    }
+
+    try (CloseableIterable<Record> writtenRecords =
+        IcebergGenerics.read(table).project(schema).useSnapshot(snapshot.snapshotId()).build()) {
+      for (Record record : writtenRecords) {
+        records.add(record);
+      }
+    }
+
+    records.sort(
+        Comparator.comparing(
+            record -> record.get(1, Long.class), Comparator.nullsFirst(Comparator.naturalOrder())));
+    return records;
+  }
+
+  protected List<Record> sortedRecordsByRowId(List<Record> records) {
+    records.sort(
+        Comparator.comparing(
+            record -> record.get(1, Long.class), Comparator.nullsFirst(Comparator.naturalOrder())));
+    return records;
+  }
+
+  protected List<Record> generateRecordsAndAppendToTable(
+      Table table, int numRecords, List<StructLike> partitions) throws IOException {
+    Schema schema = TypeUtil.join(table.schema(), rowLineageSchema);
+    List<Record> records = Lists.newArrayList();
+    Map<Integer, List<Record>> partitionToRecords = Maps.newHashMap();
+    Map<Integer, StructLike> partitionToPartitions = Maps.newHashMap();
+    for (int i = 0; i < numRecords; i++) {
+      records.add(createRecord(schema, i, (long) i, (long) i * 2));
+      int partitionKey = partitions.get(i) != null ? i : -1;
+      List<Record> existingRecords =
+          partitionToRecords.getOrDefault(partitionKey, Lists.newArrayList());
+      existingRecords.add(records.get(i));
+      partitionToRecords.put(partitionKey, existingRecords);
+    }
+
+    for (StructLike partition : partitions) {
+      if (partition != null) {
+        partitionToPartitions.put(partition.get(0, Integer.class), partition);
+      }
+    }
+
+    AppendFiles append = table.newAppend();
+    for (Map.Entry<Integer, List<Record>> entry : partitionToRecords.entrySet()) {
+      OutputFile file = Files.localOutput(temp.resolve(UUID.randomUUID().toString()).toFile());
```

Comment:
can replace all this with `FanoutDataWriter`

Reply:
Can't build a FanoutWriter without providing a FileWriterFactory implementation which the spark one isn't exposed here, but we can use PartitionMap to simplify a few things here